### ISSUE TITLE
Could we clarifies better in the semver doc PATCH releases and initial development?

### ISSUE THREAD

**Issue by camilamacedo86** [semver/semver#915]
By reading the doc https://semver.org/ is clear for me that:

- PATCH releases should not contains api changes and they are reserved to CVEs and bug fixes.
- If the project is not a stable API yet. That means the project will not have MAJOR bumps and because of it we could expected any breaking changes into the MINOR bumps.

However, the text:
> 
Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

Brings the idea that new features (api changes) would be fine in PATCH releases with ` Anything MAY change at any time.`

**Comment by camilamacedo86** [comment]
Proposed suggestion to bring clarifications to do the doc: https://github.com/semver/semver/pull/916

**Comment by ljharb** [comment]
What it means is that in `0.y.z`, everything is major, and there *is* no patch. "patch" isn't "the third number", it's "the number that indicates non-additive non-breaking changes", which doesn't exist in a 0.x version.

**Comment by camilamacedo86** [comment]
Hi @ljharb, 

The whole point of we have a convention is to have clear and define criteria/rules to define a contract to address the expectations of those that will consume the API as of those that keep the solution maintained. 

IHMO your argumentations shows inconsistent with the info provided in the doc https://semver.org/. Therefore, the https://semver.org/ shows need to properly cover the stage where the project version is < 1.0.0. Otherwise we would not have this discussion and with very different point of views.

Now, following some comments inline about your replies so far:

> What it means is that in 0.y.z, everything is major, and there is no patch. "patch" isn't "the third number", it's "the number that indicates non-additive non-breaking changes", which doesn't exist in a 0.x version.

That does not seems accurate and very contrary of the points described in this doc. If so, why we have in the doc:

<img width="897" alt="Screenshot 2023-02-09 at 08 03 30" src="https://user-images.githubusercontent.com/7708031/217752657-3cca52ae-82e4-4960-9168-5adea34e8fd3.png">

<img width="800" alt="Screenshot 2023-02-09 at 08 23 57" src="https://user-images.githubusercontent.com/7708031/217756885-387c8c83-35a3-4508-a28a-f4329723d0c5.png">

If you are right here then would not make sense the document say for projects < 1.0.0 to do releases `0.y.z` and should explicit describe to use only `0.y`. 

- Why we would describe to began with 0.1.0 because the changes will be also adding new functionalities and re not only bug fixes? 
-  What would be the point of describe that initial development also have `0.y.z` and `0.y` releases if in both scenarios you can add ANY thing? 

> Ideally you'd start the project with 1.0.0 and avoid v0 versions entirely :-)

I do not think that make sense. Because before an API and any project become stable many lifecycles and interactions are required with those users which will result in incompatible changes. I can share with you that for example many k8s projects are not stable yet.

If you think in this away then, this convention could not be applied for a lot of real scenarios. Nobody will start a project and make it stable 1.0.0 after a long period. Otherwise it would result to have a scenario like release 100.0.2 very soon. 

Also, that does not seems a good reason for we do properly define and specify what as a consumer we should expected to see in MINOR (or 0.Y.0) releases and PATCH (0.Y.Z) releases for this scenario.    

> v0 versions have no patch number.

That does not seems right. In the top of the document we have:

<img width="799" alt="Screenshot 2023-02-09 at 08 06 22" src="https://user-images.githubusercontent.com/7708031/217753212-e1c7ef0d-3e34-4959-b551-8bebffa00bd1.png">

Also, we have:

<img width="769" alt="Screenshot 2023-02-09 at 08 22 55" src="https://user-images.githubusercontent.com/7708031/217756591-565fcf70-cacd-4ce1-9b1d-957ea78b4c0f.png">

- Therefore, why the superset rules are not applied to initial development releases? What would be the reason for that? 

It is clear for me that we cannot expect from a project that is < 1.0.0 need to make a MAJOR bump releases to breaking changes because no MAJOR releases will be produced at all in this case. 

However, as I consumer of the project I would expected that the rule `PATCH version when you make backwards compatible bug fixes` would still respected and just see bug fixes in the `0.y.z` releases.

I hope that the above comments and argumentations make sense. 
Also, I would love get others input here as well.

**Comment by ljharb** [comment]
I agree there's confusion there - i think the document mistakenly conflates, at times, first/second/third for major/minor/patch, since those are only the same at v1+.

**Comment by jwdonahue** [comment]
@camilamacedo86 

Reading and writing technical specs is as much an art form as it is an engineering practice. The language can be particularly terse with respect to "normal" communications. The goal is to produce a document that can hold up to formal logic and the later requires extra attention on the part of the reader. The SemVer spec does a pretty good job of this on whole, but it is also cluttered with some non-formal language that tends to be the prime source of confusion. It's also attempting to simultaneously codify and extend a certain pre-existing convention that was also likely in use right here at github before the author checked in the original proposal.

> Why are the superset rules not applied to initial development releases? What would be the reason for that?

Because during initial development, lots of things are broken and unpredictable and the standard would not have been adopted if it did not include the prerelease loop holes. First, the 0.y.z form was a common convention prior to the writing of SemVer and second, there had to be a get of jail free card for already released software, to allow for additional prerelease cycles. If not for the allowance of these two conventions, SemVer would never have been adopted, as it would be deemed too restrictive.

There's nothing in the spec that prevents you from using the full major.minor.patch-prereleaseTag semantics for initial development. The 0.y.z prerelease form is there because some developers and probably some marketing departments didn't like the idea of their first official alpha/beta/dog-food release being very much higher than 1.0.0. In the CI/CR we live in today, best engineering practices mostly prevail, so marketing and developer predilections and superstitions are a little less dominate, and we're seeing a lot more of those higher major version numbers than in the past.  

If you start at 1.y.z-pr, you should use an informative tag of the form -a.Dev or my favorite -a.Experimental.  Note the 'a.' prefix on the tag. That's there so my Experimental tag doesn't sort higher than -Beta. Alternatively you could use -alpha.

Nothing in the SemVer spec prevents you from layering your own compatible conventions over the top of it.

> Where is it described in the documentation?

There's six parts to the SemVer doc:

1. Summary
2. Introduction
3. Semantic Version Specification (SemVer)
4. BNF Grammar
5. Why Use Semantic Versioning
6. FAQ

The spec and the BNF grammar say everything you need to know. The rest is there for for anyone initially unable to grok those two parts or unwilling to read them carefully enough. There is much that can be derived from them. Almost as important as what they say is what they don't say. The spec is not over-engineered to the point of being difficult to implement or use. If you spend some time researching through some of the discussions that have taken place here (closed as well as open), you will get a better impression of how we interpret the spec and why some of it's apparent vagaries are allowed to stand without comment in the FAQ.

 > Why can we not say that a release is 0.Y.Z is a PATCH release for a project that has not yet reached the stable version?

Why indeed?  Where in the spec does it say you cannot do that?  

There's two common conventions allowed under the initial prerelease phase, either drop the major version entirely and continue using the minor and patch as intended (0.minor.patch), or shift the meaning of the triple to the right (0.major.minor). And even though the spec doesn't say anything about the 0.m.y convention, it is certainly allowed due to the phrases "Anything MAY change at any time. The public API SHOULD NOT be considered stable". 

One problem I have with these to two initial release conventions, is that there is no formal way to distinguish between them, but then we don't have to because we have the phrases "Anything MAY change at any time. The public API SHOULD NOT be considered stable".

In this case the publishers intent is not relevant. The consumers of the initial release had better take head of the phrases "Anything MAY change at any time. The public API SHOULD NOT be considered stable".

> Why can we not say that a release 0.Y is a MINOR release for the same scenario with the intuitive caveat that also might contain incompatible changes since the project will not provide MAJOR releases in this case?

Unless I am misunderstanding the question, I think the spec says exactly that. The Y part of the triple is the Minor field and the fact that it's a prerelease version means that breakage is allowed. 

So now we come to the realization that the spec never says "MAJOR.MINOR.PATCH" (it's in the summary, not the spec), it uses the "X.Y.Z" notation everywhere and applies the labels Major (X.y.z), Minor (x.Y.z), Patch (x.y.Z) instead. Why is this? Because the original convention that was being extended with the proposed semantics commonly used the X.Y.Z pattern (there were others, but they weren't triples). In some cases X was the "major" thing that marketing was promoting as the next best thing, and the Y.Z parts were left for whatever willy-nilly thing the developers wanted to cram into them (like dates!). In other cases it was X => Year, Y => Month, etc.  All of those other conventions are still quite commonly used, but many of them already have their own set of implicit or implied semantics.

**Comment by jwdonahue** [comment]
I should note that I have also seen the 0.YYYYM.DRR convention is used internally in many places.  You can't have leading zeros in the triple, so the year and revision parts must be fixed length so you can drop the leading zeros for month and day fields as needed and still parse the thing in your head.

I don't see any reason why you couldn't release those to early adopters. The semantics are obvious when you see them in the wild and the Y and Z fields will get reset when you finally get around to bumping the major version. Since both prerelease forms absolve the publisher from semantic correctness with regard to the triple values, they lose their intended meanings from the perspective of the consumer anyway, so why not lay in some useful information instead?

You may encounter some tooling that won't accept large numeric field values because the implementers either misinterpreted the spec to mean integer in whatever their programming language was (8, 16, 32, 64 bits, etc) or they were just to lazy to figure out that an integer string without leading zeros will sort correctly if you take account of the string length (simple algorithm takes less time to sort than converting to int and then comparing, in the vast majority of real world version strings).

**Comment by steveklabnik** [comment]
For whatever it's worth, I see this line as the worst line in the semver spec

> Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

And would like to see it removed. Nobody who implements semver actually respects this; they all implement the same ranges as everyone else, some of which have semantic meaning for the relationships between x=0 versions.

I just have not had the energy to try and get it done. Maybe I will.

**Comment by ljharb** [comment]
@steveklabnik do you mean to enshrine the way npm treats v0 versions? If that's the case and you'd accept the PR, I'm happy to take a crack at authoring it.

**Comment by steveklabnik** [comment]
I think "the way npm treats v0 versions" is too broad for me to know what specifically you mean; I'm referring to the "the ^ matcher treats y as x and z as y when x = 0" semantics.

But part of the reason that I haven't done it, and maybe you're interested in this too, is that IMHO the spec is poorly organized. That's nobody's fault, really. But at the *very* least, the "what versions mean" bits and the "how versions relate to each other" bits should be separated out cleanly. Part of the issue here is that https://github.com/semver/semver/pull/584 has still not landed; I think ideally this would all be under a "how versions relate to each other" bit.

But that's not all fully thought through, and a bit rambly. Needs more work, heh.

**Comment by ljharb** [comment]
Gotcha, I think that aligns :-) I'll see if I can find some time this week. While I'd love to see #584 land, that seems like a larger change and i don't think this needs to wait on it.

**Comment by steveklabnik** [comment]
Cool cool. I agree that it's not blocking, for sure.

**Comment by jwdonahue** [comment]
Reducing to only one prerelease form would help. But that's a breaking change, so how do you disambiguate between SemVer 2 and 3?  I strongly recommend that you resolve that problem first.

**Comment by steveklabnik** [comment]
> Reducing to only one prerelease form would help.

That is not what I would advocate. Pre-release versions are still pre-release versions.

**Comment by jwdonahue** [comment]
@steveklabnik I meant that the 0.y.z form of prerelease is redundant.  I had thought you were in agreement with that, or were you objecting to the "Anything MAY change at any time. The public API SHOULD NOT be considered stable" phrasing of the [4th clause](https://semver.org/spec/v2.0.0.html#spec-item-4)?

**Comment by steveklabnik** [comment]
>  I had thought you were in agreement with that

I am not. People use this, massively, in the real world. There's no advantage to getting rid of this. In fact, in some ways, the point is the opposite: people don't necessarily view x=0 as a prerelease.

> were you objecting to the "Anything MAY change at any time. The public API SHOULD NOT be considered stable" phrasing of the [4th clause](https://semver.org/spec/v2.0.0.html#spec-item-4)?

That is correct. Things like the `^` operator define stricter semantics than `MAY`, which still technically fits the spec, but the split between real-world usage and the 4th clause causes endless online arguments. The 4th clause should be brought in line with real-world usage.

**Comment by ljharb** [comment]
Yes, that's the part we discussed changing. I'll make a PR for it this week.

**Comment by jwdonahue** [comment]
> Things like the ^ operator define stricter semantics than MAY

I am not familiar with how that operator is applied to version ranges in whatever tool chain(s) you are referring to.  Stricter in what way?

**Comment by ljharb** [comment]
@jwdonahue in npm, `^` includes nonbreaking changes. so in `^x.y.z`, the y and z may change. Similarly, in `v0.x.y` and `v0.0.x`, only the y and z may change - because the x is always the "major", for breaking changes.

**Comment by steveklabnik** [comment]
(not only npm, but cargo, and iirc bundler with `~>`.

**Comment by ljharb** [comment]
@steveklabnik filed #923, lmk your thoughts!

**Comment by jwdonahue** [comment]
I just stumbled across a related comment of old: https://github.com/semver/semver/pull/127#issuecomment-47964783

We now have multiple PR's that are related to this topic.

**Comment by ljharb** [comment]
@jwdonahue npm isn't semver v2 compliant, because npm existed before the current version of the semver spec. #932 and #584, however, would change the semver spec so that it fully matches npm.

**Comment by jwdonahue** [comment]
> @jwdonahue npm isn't semver v2 compliant, because npm existed before the current version of the semver spec. https://github.com/semver/semver/issues/932 and https://github.com/semver/semver/pull/584, however, would change the semver spec so that it fully matches npm

Are you saying that NPM cannot be used in its current form in a SemVer compliant way? Does it not have a -SemVer mode? That would be a sad situation to be in. I am sure there are other packaging tools out there who's owners wish they could bend the SemVer spec to validate their current product behaviors. 

This is one reason why I have always argued that the spec should steer clear of the range specification and dependency tree problems. SemVer as it is, is a simple convention to follow. Adding range spec behavior and dependency tracking to it, will constrain innovation and severely complicate the spec.  Better to layer-on whatever flavor of range spec you prefer to use in a separate document.  In the case of the ^ operator, it sounds like you also have non SemVer compliant semantics of your own.

I don't see any problem with the various packaging schemes applying their preferred semantics, provided they are clear that theirs are not SemVer semantics. SemVer is a specific kind of semantic versioning that is very simple, easy to implement and programming language/tool-chain agnostic. Other than specifying how to sort a list of SemVer version strings, it is entirely agnostic on the subject of range specifications and other implementation details.  

Tool owners and development communities may choose to support SemVer or any other form of semantic versioning that they see fit. Claiming to use a form of semantic versioning is not the same thing as claiming SemVer compliance. A tool need not promote any of the semantic forms it supports. I know there are many that predate SemVer which also currently support it while allowing for other forms to be used.  It sounds like NPM promotes another form by implication of its range operator specifications.

**Comment by ljharb** [comment]
I'm saying that "what semver is" changed after npm had already cemented its semantics, and no, there couldn't possibly be any mode to give it different semantics. Your argument already doesn't work because this spec already changed (widening what changes are expected in v0) in a way that conflicted with existing implementations. #932 fixes that.

**Comment by ljharb** [comment]
The purpose of specs is to document reality, not to dictate it. The spec should match what major implementations are already doing.

**Comment by jwdonahue** [comment]
Then we'd need 10 of them.

**Comment by ljharb** [comment]
Or, we'd need one with sufficient minimal union semantics that all 10 can be considered compliant.

**Comment by jwdonahue** [comment]
There are literally thousands of bespoke tools out there that implement SemVer 2.0.0 as it is.  Why break it for a handful of tools that are getting by just fine without it?

**Comment by jwdonahue** [comment]
> The purpose of specs is to document reality, not to dictate it. The spec should match what major implementations are already doing.

No, that's what your product docs are for. This spec was intended by mojombo to be a proposal for a specific form of semantic versioning. I think NPM predates the spec and since https://semver.org/spec/v1.0.0-beta.html, the language regarding the 0.y.z form has been very clear that "The public API should not be considered stable" and was then changed to "Anything may change at any time. The public API should not be considered stable" in the [v1 spec.](https://semver.org/spec/v1.0.0.html).

I don't see anything in [Preston's LinkedIn profile](https://www.linkedin.com/in/mojombo/details/experience/) indicating that he's ever had anything to do with NPM. I have always suspected that he wrote it in response to dependency issues experienced during the development of GitHub, caused by the subjective use of major.minor.patch versioning that was still common at the time. It's obvious the proposal simply layered tighter semantics on meanings of each of those fields.

I would add that most specifications are written prior to implementation. The general purpose of writing specs is to make it clear how the implementation should behave or what it should look like. I would also add that there are packaging and other versioning related tools out there that do adhere to the spec as written (Nuget for instance).

**Comment by ljharb** [comment]
Originalism isn't really useful or realistic - the value of the spec now is that it's what people look to when they want to mirror existing implementations, and if it doesn't match those, it's not doing its job.

**Comment by jwdonahue** [comment]
There are plenty of implementations that hone to the current spec. Many that are mostly compatible, plus support for other semantics as well. Tweaking the spec to match NPM behavior will break it and render many tools incompatible. From that point on, the spec won't be relevant to anybody but the NPM community and yet another fork will follow.

I think the main variance between all of the tool chains is in their range specifiers and selectors used for defining dependencies. It's already pretty obvious that not even the tools represented by the maintainers can agree how those should operate and they may represent a majority of the package repositories, they are a minority in terms of CI/CD tooling however. There are literally thousands of other tools, some of them representing large vertical segments that reference this spec.

I think the whole point of having a very narrowly defined spec is to clarify the gist of semantic versioning (not really invented by mojombo btw) and leave some space to innovate.

**Comment by jorenbroekema** [comment]
As far as I understand, NPM's implementation of the semver spec and its ambiguity around 0.y.z and 0.0.z versions is as follows:
> For x, y and z, if the preceeding digit is 0, and current is a non-0 digit, then current is the major.

That means:
0.y.z -> y = major, z = minor
0.0.z -> z = major

Maybe this can be worded better but I'm just curious if that by itself would fit the spirit of the semver spec while not making a bunch of implementation considered non-compliant. At least for me this makes a lot of sense conceptually, NPM already implements it this way.. do we know how far other semver implementors stray from this principle?

What's really annoying, as a package maintainer, is getting constant complaints about "breaking changes in minors/patches" and not being able to point them to either the semver spec or the NPM semantic versioning docs, because neither of those documents actually get specific about this behavior. So all I can do is say "sorry that's how NPM versions works, you're wrong and you'll just have to trust me as there is no explicit documentation on this convention"

**Comment by steveklabnik** [comment]
For what it's worth, I too have been frustrated by the lack of clarity on this topic. I had the first conversation with someone about it  was pre-pandemic, but sadly other things were higher priority, and then I have had pretty extreme OSS burnout. I still hold out hope that maybe I can get it across the line someday, but I do not know when that might be.

**Comment by jwdonahue** [comment]
The more I think about this, the more I think this is just a means to allow 2 or 1 digit version numbers and still call it SemVer, so why not expand it to 4 or more digits as well?

The SemVer version disambiguation problem comes to mind.
